### PR TITLE
GH-93678: reduce boilerplate and code repetition in the compiler

### PR DESCRIPTION
**Pull request by iritkatriel**
It might be easier to review the individual commits separately.

The code generation part of the compiler has several functions which are slight variations of each other. Part of this is due to the evolution of the code when accurate line number information was added.

This PR tidies this up to simplify the code. Having just one basicblock_addop function that knows how to handle all the HAS_ARG, !HAS_ARG and JUMP cases is a step towards being able to feed bytecode instruction sequences and turn them into basicblocks (without having to differentiate between the three types).

### CODE REVIEW

**Comment by Fidget-Spinner** [comment]
Sorry but I'm clueless when it comes to the compiler, so I can't give a review.

File: Python/compile.c [hunk 1]
```diff
@@ -1269,48 +1292,46 @@ compiler_use_new_implicit_block_if_needed(struct compiler *c)
...
-basicblock_addop_line(basicblock *b, int opcode, int line,
-                      int end_line, int col_offset, int end_col_offset)
+basicblock_addop(basicblock *b, int opcode, int oparg,
+                 basicblock *target, struct location loc)
 {
     assert(IS_WITHIN_OPCODE_RANGE(opcode));
     assert(!IS_ASSEMBLER_OPCODE(opcode));
-    assert(!HAS_ARG(opcode) || IS_ARTIFICIAL(opcode));
+    assert(HAS_ARG(opcode) || oparg == 0);
+    assert(0 <= oparg && oparg <= 2147483647);
```
Comment (markshannon) [inline Python/compile.c]: We assume that it is OK to convert code unit offsets to byte offsets, and there may be other cases where we do unchecked doubling. So best to limit to `1 << 30`.

**Review by markshannon — COMMENTED** [review]
Seems like a nice improvement in the architecture of the compiler.

**Review by brandtbucher — APPROVED** [review]
Thanks, this looks nice.

One thing I noticed is that the new design creates lots of temporary `location` structs (at least one for every instruction emitted by the frontend, I think). Could we instead keep just one struct on the compiler unit, and maybe have a static "no location" struct, and just pass references to those into `basicblock_addop` instead?

**Comment by iritkatriel** [comment]
> Thanks, this looks nice.
> 
> One thing I noticed is that the new design creates lots of temporary `location` structs (at least one for every instruction emitted by the frontend, I think). Could we instead keep just one struct on the compiler unit, and maybe have a static "no location" struct, and just pass references to those into `basicblock_addop` instead?

Yeah I guess we could pass them by reference. 

I'm now working on replacing the four fields by the struct in the instr and compiler unit structs, will push that in a bit.

File: Python/compile.c [hunk 2]
```diff
@@ -948,24 +960,15 @@ basicblock_next_instr(basicblock *b)
...
-#define SET_LOC(c, x)                           \
-    (c)->u->u_lineno = (x)->lineno;             \
-    (c)->u->u_col_offset = (x)->col_offset;     \
-    (c)->u->u_end_lineno = (x)->end_lineno;     \
-    (c)->u->u_end_col_offset = (x)->end_col_offset;
+#define SET_LOC(c, x)                             \
+    (c)->u->u_loc = LOCATION((x)->lineno,         \
+                             (x)->end_lineno,     \
+                             (x)->col_offset,     \
+                             (x)->end_col_offset)
```
Comment (iritkatriel) [inline Python/compile.c]: Is this is slower than the older version? Or will the compiler know what to do?

**Review by markshannon — APPROVED** [review]
Nice improvement in clarity.

File: Python/compile.c [hunk 3]
```diff
@@ -144,17 +149,26 @@
...
+    int lineno;
+    int end_lineno;
+    int col_offset;
+    int end_col_offset;
+};
+
+#define LOCATION(LNO, END_LNO, COL, END_COL) \
+    ((const struct location){(LNO), (END_LNO), (COL), (END_COL)})
+
+#define NO_LOCATION {-1, -1, -1, -1}
```
Comment (sweeneyde) [inline Python/compile.c]: Could this be
```c
static struct location NO_LOCATION = {-1, -1, -1, -1};
```
(without the `(struct location)` cast of 5e06cad8cf2d4023ff97cd6eefef5e2f4166eba2)?

**Review by sweeneyde — APPROVED** [review]
100 lines shorter 🎉

File: Python/compile.c (hunk 3 above)
Comment (iritkatriel) [reply Python/compile.c]: That might work, it will remove another 6 lines or so!

**Pull request merged**